hawkular-wildfly-monitor/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java

```diff
@@ -0,0 +1,444 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.hawkular.agent.monitor.storage;
+
+import static org.hawkular.inventory.api.Relationships.WellKnown.incorporates;
+
+import java.io.Reader;
+import java.io.StringReader;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+
+import org.hawkular.agent.monitor.api.InventoryStorage;
+import org.hawkular.agent.monitor.extension.MonitorServiceConfiguration;
+import org.hawkular.agent.monitor.extension.MonitorServiceConfiguration.StorageAdapter;
+import org.hawkular.agent.monitor.inventory.AvailInstance;
+import org.hawkular.agent.monitor.inventory.AvailType;
+import org.hawkular.agent.monitor.inventory.ID;
+import org.hawkular.agent.monitor.inventory.MeasurementInstance;
+import org.hawkular.agent.monitor.inventory.MeasurementType;
+import org.hawkular.agent.monitor.inventory.MetricInstance;
+import org.hawkular.agent.monitor.inventory.MetricType;
+import org.hawkular.agent.monitor.inventory.NamedObject;
+import org.hawkular.agent.monitor.inventory.Resource;
+import org.hawkular.agent.monitor.inventory.ResourceConfigurationPropertyInstance;
+import org.hawkular.agent.monitor.inventory.ResourceType;
+import org.hawkular.agent.monitor.log.AgentLoggers;
+import org.hawkular.agent.monitor.log.MsgLogger;
+import org.hawkular.agent.monitor.service.ServerIdentifiers;
+import org.hawkular.agent.monitor.service.ThreadFactoryGenerator;
+import org.hawkular.agent.monitor.service.Util;
+import org.hawkular.inventory.api.Relationships.Direction;
+import org.hawkular.inventory.api.Resources;
+import org.hawkular.inventory.api.model.AbstractElement;
+import org.hawkular.inventory.api.model.CanonicalPath;
+import org.hawkular.inventory.api.model.DataEntity;
+import org.hawkular.inventory.api.model.Entity;
+import org.hawkular.inventory.api.model.Metric;
+import org.hawkular.inventory.api.model.MetricDataType;
+import org.hawkular.inventory.api.model.MetricUnit;
+import org.hawkular.inventory.api.model.Relationship;
+import org.hawkular.inventory.api.model.StructuredData;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.squareup.okhttp.Request;
+import com.squareup.okhttp.Response;
+
+/**
+ * @author <a href="https://github.com/ppalaga">Peter Palaga</a>
+ */
+public class AsyncInventoryStorage implements InventoryStorage {
+    private static class BulkPayloadBuilder {
```

Comment:
I really would like to see some good javadoc on these methods explaining what they are doing.